File: packages/sdk/src/domains/rpc.ts
/**
 * The typed request/event bridge every domain is built on.
 *
 * A domain never touches {@link GatewayTransport} directly. It calls through this thin layer, which
 * pairs an opcode with the generated `encodeX`/`decodeX` for that opcode's payload and response, so a
 * domain method reads as one line — `rpc.call(OP.SYNC, encodeSyncRequest, decodeSyncResponse, req)` —
 * and a mismatched struct is a compile error rather than a wire fault.
 *
 * # Why event decoding is wrapped
 *
 * The transport fans a server event out to its listeners by calling them inline in its receive loop,
 * and it does not catch what a listener throws (see `GatewayTransport`'s dispatch). A listener that
 * throws — because a single event decoded badly, or an application handler had a bug — would escape
 * into the socket's `onmessage` and take down the whole receive path, losing every subsequent frame.
 * So {@link Rpc.on} decodes and dispatches inside a `try`/`catch`: one malformed or mishandled event
 * is routed to an optional error sink and dropped, and the connection keeps delivering.
 */

import { OP, CODE } from '@migo/protocol';
import type { Frame } from '@migo/wire';

import { decodeBody, encodeBody } from '../codec.js';
import type { BodyDecoder, BodyEncoder } from '../codec.js';
import { RemoteError } from '../errors.js';
import type { GatewayTransport } from '../transport.js';

/**
 * The read opcodes a RATE_LIMITED reply retries once against. These are the listing/reporting
 * surfaces a UI fires in bursts as sections open — exactly the traffic the server's advice
 * (`retry_after_ms`) can absorb — and they are all safe to re-issue: none of them mutates
 * anything. Everything else (a message send, a gift, a friend action) surfaces its error, because
 * an automatic second send is how a message arrives twice.
 */
const RETRYABLE_READS: ReadonlySet<number> = new Set([
  OP.CONVERSATION_LIST,
  OP.SYNC,
  OP.ROOM_LIST,
  OP.ROOM_ROSTER,
  OP.RELATIONSHIP_LIST,
  OP.PROFILE_FETCH,
  OP.NOTIFICATION_LIST,
  OP.SEARCH,
  OP.SUGGESTIONS,
  OP.BALANCE_FETCH,
  OP.GIFT_CATALOGUE,
  OP.LEDGER_HISTORY,
  OP.PROGRESSION,
  OP.BADGES,
  OP.LEADERBOARD,
]);

/** The largest backoff a server's advice can command; anything larger is a policy worth surfacing. */
const MAX_RETRY_DELAY_MS = 10_000;

/** Notified when an inbound event fails to decode, or its handler throws. Never fatal. */
export type EventErrorHandler = (opcode: number, cause: unknown) => void;

/** A decoded server event and the frame it arrived on. */
export type EventHandler<Event> = (event: Event, frame: Frame) => void;

/**
 * The request/notify/subscribe surface shared by all domains.
 *
 * Holds no protocol knowledge of its own — the opcode and the codecs are always passed in by the
 * domain — so it stays a mechanical adapter over the transport.
 */
export class Rpc {
  readonly #transport: GatewayTransport;
  readonly #onEventError: EventErrorHandler | undefined;

  constructor(transport: GatewayTransport, onEventError?: EventErrorHandler) {
    this.#transport = transport;
    this.#onEventError = onEventError;
  }

  /**
   * Sends a request and decodes its reply.
   *
   * An ERROR-flagged reply has already rejected as a {@link RemoteError} inside the transport, and a
   * missing reply as a {@link TimeoutError}; this only runs on the success path, so the decode is
   * always against the opcode's declared response type.
   */
  async call<Request, Response>(
    opcode: number,
    encode: BodyEncoder<Request>,
    decode: BodyDecoder<Response>,
    request: Request,
  ): Promise<Response> {
    try {
      const frame = await this.#transport.request(opcode, encodeBody(encode, request));
      return decodeBody(decode, frame.payload);
    } catch (cause) {
      // The server answered a read with "too many, wait N ms" and the client obeys — once. This
      // is the official client's half of the rate-limit contract: it never retries a mutation,
      // and it caps the wait so a hostile or misconfigured node cannot stall a UI for minutes.
      if (
        RETRYABLE_READS.has(opcode) &&
        cause instanceof RemoteError &&
        cause.code === CODE.RATE_LIMITED &&
        cause.retryAfterMs !== undefined
      ) {
        const wait = Math.min(Math.max(0, cause.retryAfterMs), MAX_RETRY_DELAY_MS);
        await new Promise((resolve) => setTimeout(resolve, wait));
        const frame = await this.#transport.request(opcode, encodeBody(encode, request));
        return decodeBody(decode, frame.payload);
      }
      throw cause;
    }
  }

  /** Sends a fire-and-forget frame the protocol gives no reply to (TYPING, MESSAGE_RECEIPT). */
  async notify<Request>(
    opcode: number,
    encode: BodyEncoder<Request>,
    request: Request,
  ): Promise<void> {
    await this.#transport.notify(opcode, encodeBody(encode, request));
  }

  /**
   * Subscribes to a server event opcode, decoding each frame before the handler sees it.
   *
   * A decode failure or a throw from `handler` is delivered to the error sink and swallowed, so it
   * never propagates into the transport's receive loop. Returns the transport's unsubscribe function.
   */
  on<Event>(opcode: number, decode: BodyDecoder<Event>, handler: EventHandler<Event>): () => void {
    return this.#transport.subscribe(opcode, (payload, frame) => {
      let event: Event;
      try {
        event = decodeBody(decode, payload);
      } catch (cause) {
        this.#onEventError?.(opcode, cause);
        return;
      }
      try {
        handler(event, frame);
      } catch (cause) {
        this.#onEventError?.(opcode, cause);
      }
    });
  }
}

File: clients/android/core/src/main/kotlin/com/migo/core/domain/Rpc.kt
package com.migo.core.domain

import com.migo.core.net.Gateway
import com.migo.core.net.GatewayError
import com.migo.core.net.RealtimeTransport
import com.migo.core.wire.Frame
import com.migo.core.wire.Reader
import com.migo.core.wire.Writer
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

/**
 * The typed request/event bridge every domain is built on.
 *
 * A domain never touches [Gateway] directly. It calls through this thin layer, which pairs an opcode
 * with the generated `encode(Writer)` and `decode(Reader)` for that opcode's payload, so a domain
 * method reads as one line and a mismatched struct is a compile error rather than a wire fault.
 * Mirrors `packages/sdk/src/domains/rpc.ts`.
 *
 * # Why there is a pump here and not in the transport
 *
 * The web SDK's transport dispatches events to listeners from inside its own receive loop. This
 * client cannot: OkHttp delivers frames on its own thread, and [Gateway] does the sequence
 * bookkeeping synchronously with arrival because the ACK watermark has to be exact. Running
 * application handlers on that thread would put arbitrary UI code in the path of the frame counter.
 * So [Gateway] hands uncorrelated frames to a channel and [deliver] is the seam: one coroutine drains
 * that channel and fans each frame out to the domains. Nothing else may consume [Gateway.inbound] --
 * a channel with two readers would deliver each event to one of them at random.
 *
 * [deliver] must be running for any `on` subscription to fire, and it must keep running for the whole
 * connection: the channel is unbounded, so a client that stopped draining would hold every broadcast
 * it ever received in memory rather than dropping them.
 *
 * # Why a handler cannot break the connection
 *
 * [deliver] decodes and dispatches inside a `try`/`catch`. One malformed event, or one application
 * handler with a bug, is routed to [EventErrorHandler] and dropped; the pump keeps delivering. The
 * alternative is a single throw from a UI callback ending the receive loop and silently costing the
 * user every subsequent message.
 */

/** Notified when an inbound event fails to decode, or its handler throws. Never fatal. */
typealias EventErrorHandler = (opcode: Long, cause: Throwable) -> Unit

/** A registration that can be undone. Returned by every `on` method in every domain. */
fun interface Subscription {
    /** Removes the handler. Idempotent: cancelling twice is not an error. */
    fun cancel()
}

/**
 * The request/notify/subscribe surface shared by all domains.
 *
 * Holds no protocol knowledge of its own -- the opcode and the codecs are always passed in by the
 * domain -- so it stays a mechanical adapter over the gateway.
 */
class Rpc(
    private val gateway: RealtimeTransport,
    private val onEventError: EventErrorHandler? = null,
) {
    /**
     * Guards the subscriber table.
     *
     * A plain lock rather than a [kotlinx.coroutines.sync.Mutex]: registering a handler is a map
     * insert, callers do it from arbitrary threads while composing a screen, and making `on` a
     * suspending function would push that into every caller for no benefit. Handlers are dispatched
     * outside the lock -- see [dispatch].
     */
    private val lock = ReentrantLock()

    private val subscribers = HashMap<Long, MutableList<Entry>>()

    /**
     * Sends a request and decodes its reply.
     *
     * An ERROR-flagged reply has already become [GatewayError.Refused] inside [Gateway.request], so
     * this only runs on the success path and the decode is always against the opcode's declared
     * response type.
     *
     * A reply this build cannot parse is [GatewayError.Malformed] rather than the underlying
     * [com.migo.core.wire.WireError]: which field ran short is a fact about bytes from the network,
     * and a caller can do nothing with it that it would not also do with "the server's answer was
     * unreadable".
     */
    suspend fun <Res> call(opcode: Long, encode: (Writer) -> Unit, decode: (Reader) -> Res): Res {
        val frame = gateway.request(opcode, encode)
        return decodePayload(frame, decode)
    }

    /**
     * Sends a fire-and-forget frame the protocol gives no reply to (TYPING, MESSAGE_RECEIPT).
     *
     * Correlation zero, which is what "not a reply to anything, and no reply expected" is on the
     * wire. Using a fresh correlation id instead would leave the server free to answer, and this
     * client with no waiter to answer to.
     */
    suspend fun notify(opcode: Long, encode: (Writer) -> Unit) {
        gateway.send(opcode, NOT_A_REPLY, encode)
    }

    /**
     * Subscribes to a server event opcode, decoding each frame before the handler sees it.
     *
     * Several handlers may share an opcode and all of them fire. A decode failure or a throw from
     * [handler] is delivered to the error sink and swallowed, so it never reaches [deliver]'s loop.
     */
    fun <Event> on(
        opcode: Long,
        decode: (Reader) -> Event,
        handler: (Event, Frame) -> Unit,
    ): Subscription {
        val entry = Entry { frame ->
            val event = try {
                decodePayload(frame, decode)
            } catch (cause: Throwable) {
                onEventError?.invoke(opcode, cause)
                return@Entry
            }
            try {
                handler(event, frame)
            } catch (cause: Throwable) {
                onEventError?.invoke(opcode, cause)
            }
        }
        lock.withLock { subscribers.getOrPut(opcode) { ArrayList() }.add(entry) }
        return Subscription {
            lock.withLock {
                val list = subscribers[opcode]
                if (list != null) {
                    list.remove(entry)
                    if (list.isEmpty()) subscribers.remove(opcode)
                }
            }
        }
    }

    /**
     * Drains the gateway's inbound frames until the connection ends. Suspends for the whole session.
     *
     * Returns normally when the socket was closed cleanly and throws the [GatewayError] the
     * connection died of otherwise, because that is the signal a reconnect policy needs and the pump
     * is where it surfaces. The caller launches this once per connection and treats its completion as
     * "this connection is over".
     */
    suspend fun deliver() {
        for (message in gateway.inbound) {
            dispatch(message.frame)
        }
    }

    /**
     * Hands one frame to every handler registered for its opcode.
     *
     * The subscriber list is copied under the lock and the handlers run outside it. Holding the lock
     * across dispatch would mean an application handler that registered another handler deadlocked
     * the connection, and reentrancy is not something a UI callback should have to know about.
     *
     * An ERROR-flagged frame with no waiter is reported rather than decoded. It is the server
     * refusing something whose reply nobody is holding any more -- a request whose caller was
     * cancelled, typically -- and its payload is an `Error`, not the struct this opcode declares. A
     * client that fed it to the opcode's decoder would report a malformed event and lose the reason.
     */
    private fun dispatch(frame: Frame) {
        val opcode = frame.header.opcode
        if (Gateway.isError(frame)) {
            onEventError?.invoke(opcode, Gateway.refusal(frame))
            return
        }
        val handlers = lock.withLock { subscribers[opcode]?.toList() } ?: return
        for (entry in handlers) {
            entry.handle(frame)
        }
    }

    /**
     * Decodes a frame's payload with [decode].
     *
     * [Reader.finish] is deliberately not called. A newer server may append an optional field to a
     * struct this build knows, and the generated decoders skip unknown optional fields by length --
     * that is the whole forward-compatibility mechanism. Asserting the payload was fully consumed
     * would turn a compatible addition into a client that refuses every frame of that opcode. This
     * matches [Gateway]'s own WELCOME decode.
     */
    private fun <Res> decodePayload(frame: Frame, decode: (Reader) -> Res): Res = try {
        decode(Reader(frame.payload))
    } catch (_: Exception) {
        throw GatewayError.Malformed
    }

    private companion object {
        /** The correlation for a frame that is neither a request nor a reply. */
        const val NOT_A_REPLY = 0L
    }
}

/**
 * One registered handler, wrapped so removal is by identity.
 *
 * A bare `(Frame) -> Unit` in the list would be removed by equality, and two subscriptions that
 * happened to capture nothing could compare equal -- unsubscribing one would then remove the other.
 */
private class Entry(val handle: (Frame) -> Unit)
